app/assets/javascripts/katello/hosts/host_and_hostgroup_edit.js

```diff
@@ -1,100 +1,172 @@
-    toggle_installation_medium();
-  });
-  $('#hostgroup_operatingsystem_id').live('change', function() {
-    toggle_installation_medium();
-  });
-}
-
-function update_environment_label(item) {
-    var lifecycle_env = $(item).val().trim();
-    if (lifecycle_env.length > 0) {
-       $("#host_environment_id").parent().parent().find('label').text('Content View');
-       $("#hostgroup_environment_id").parent().parent().find('label').text('Content View');
+var KT = KT ? KT : {};
+KT.hosts = {};
+
+$(document).on('ContentLoad', function(){
+    KT.hosts.onKatelloHostEditLoad();
+
+    $("#hostgroup_lifecycle_environment_id").change(KT.hosts.fetchContentViews);
+    $("#host_lifecycle_environment_id").change(KT.hosts.fetchContentViews);
+
+    $("#hostgroup_content_view_id").change(KT.hosts.contentViewSelected);
+    $("#host_content_view_id").change(KT.hosts.contentViewSelected);
+    $("#reset_puppet_environment").click(function() {
+        KT.hosts.getPuppetEnvironmentSelect().data('content_puppet_match', 'true');
+        KT.hosts.setDefaultPuppetEnvironment(KT.hosts.getSelectedContentView(), KT.hosts.getSelectedEnvironment());
+    });
+
+});
+
+KT.hosts.fetchContentViews = function () {
+    var select = KT.hosts.getContentViewSelect();
+    var envId = KT.hosts.getSelectedEnvironment();
+    select.find('option').remove();
+    if (envId) {
+        KT.hosts.signalContentViewFetch(true);
+        $.get('/katello/api/v2/content_views/', {environment_id: envId, full_result: true}, function (data) {
+            select.find('option').remove();
+            select.append($("<option />"));
+            $.each(data.results, function(index, view) {
+                select.append($("<option />").val(view.id).text(view.name));
+            });
+            KT.hosts.signalContentViewFetch(false);
+        });
+    }
+};
+
+KT.hosts.signalContentViewFetch = function(fetching) {
+    var select = KT.hosts.getContentViewSelect(),
+        //parent = select.parent(),
+        spinner = $('<img>').attr('src', '/assets/spinner.gif'),
+        spinner_id = "content_view_spinner";
+
+    if(fetching) {
+        select.hide();
+        $(spinner).attr('id', spinner_id).insertAfter(select);
     } else {
-       $("#host_environment_id").parent().parent().find('label').text('Puppet Environment');
-       $("#hostgroup_environment_id").parent().parent().find('label').text('Puppet Environment');
+        select.show();
+        $('#' + spinner_id).remove();
+    }
+};
+
+
+KT.hosts.contentViewSelected = function() {
+    if (KT.hosts.getPuppetEnvironmentSelect().data('content_puppet_match')) {
+        KT.hosts.setDefaultPuppetEnvironment(KT.hosts.getSelectedContentView(), KT.hosts.getSelectedEnvironment());
+    }
+};
+
+KT.hosts.setDefaultPuppetEnvironment = function(view_id, env_id) {
+    if (view_id && env_id) {
+        $.get('/hosts/puppet_environment_for_content_view', {content_view_id: view_id, lifecycle_environment_id: env_id}, function (data) {
```

Comment:
Hmm, did this change since the last time I reviewed the PR? This route doesn't seem to work, https://katello-dev.bitbin.de/hosts/puppet_environment_for_content_view?content_view_id=2&lifecycle_environment_id=1
 hits the Foreman route looking for a host called `puppet_environment_for_content_view`

```
not found: Couldn't find Host::Managed with id=puppet_environment_for_content_view [WHERE "hosts"."organization_id" IN (1) AND "hosts"."type" IN ('Host::Managed')]
  Rendered common/404.html.erb within layouts/application (1.3ms)
  Rendered layouts/base.html.erb (25.7ms)
```